### DISCUSSION TITLE
How to use testkit to improve Mill

### DISCUSSION THREAD

**Discussion by nafg** [build-server-protocol/build-server-protocol#689] [General]
I've been having some issues with IntelliJ and Mill, and on Discord @lefou suggested adding a test suite based on bsp testkit could help, but I don't know anything about it. How would one go about doing that?

**Comment by lefou** [comment]
The discussion was about Mill BSP and how to improve Mill BSP. Mill has a pretty good test coverage but the BSP component does not, so the best start to improve BSP is to add test coverage which additionally  enables a developer to reproduce and fix reported issues.

The [build-server-protocol project](https://github.com/build-server-protocol/build-server-protocol) has a `testkit`. There is a also successor [`bsp-testkit2`](https://github.com/build-server-protocol/bsp-testkit2), but it's not documented and there isn't any feedback, which makes adding it to Mill non-trival.

* https://github.com/build-server-protocol/bsp-testkit2/issues/1

So, if anybody want to improvement to Mill BSP, contributing a test setup would IMHO to most helpful task.

**Comment by jastice** [comment]
cc @anthonygrod as he is working on Mill IntelliJ integration

**Reply by AnthonyGrod** [reply]
Hello, currently I'm working on integrating Mill over IJ-BSP plugin (https://github.com/JetBrains/intellij-bsp). At the moment I'm focusing on very basic functionalities first but I'll be happy to help when I will finish the basic support.

**Reply by nafg** [reply]
Is that what current BSP import uses or is that meant to replace the current functionality?

**Comment by jastice** [comment]
> Is that what current BSP import uses or is that meant to replace the current functionality?

Current functionality in Scala plugin will be superseded by the dedicated BSP plugin. Mill will be the first use case.

As for using the testkit, maybe @agluszak can give some pointers.
